7.0 secure pvaccess

george-mcintyre opened the pull request:
### Add support for METHOD, AUTHORITY, and PROTOCOL in Access Security

- Extended Access Security Group (ASG) rules with METHOD, AUTHORITY, and PROTOCOL predicates
- Support for the new RPC permission type
- Updated related tests and configurations to validate the new functionalities
- Add new APIs to transmit new client IDENTITY to security system

Added `METHOD`, `AUTHORITY`, and `PROTOCOL` as follows:

```
UAG(bar) {boss}
UAG(foo) {testing}
UAG(ops) {geek}

AUTHORITY(AUTH_EPICS_ROOT, "EPICS Org Root CA") {
	AUTHORITY(AUTH_INTERMEDIATE_CA, "Intermediate CA") {
		AUTHORITY(AUTH_ORNL_CA, "ORNL Org CA")
	}
	AUTHORITY(AUTH_UNRELATED_CA, "Unrelated CA")
}

ASG(DEFAULT) {
	RULE(0, NONE)
}

ASG(ro) {
	RULE(0, NONE)
	RULE(1, READ) {
		UAG(foo,ops)
		METHOD(\"ca\")
		PROTOCOL(\"TCP\")
	}
}

ASG(rw) {
	RULE(0, NONE)
	RULE(1, WRITE, TRAPWRITE) {
		UAG(foo)
		METHOD("x509")
		AUTHORITY(AUTH_UNRELATED_CA)
	}
}

ASG(rwx) {
	RULE(0, NONE)
	RULE(1, RPC) {
		UAG(bar)
		METHOD("x509","ignored","ignored_too")
		AUTHORITY(AUTH_UNRELATED_CA, AUTH_ORNL_CA)
		PROTOCOL("TLS")
	}
}
```
AUTHORITY Definitions:
- AUTHORITY 
  - definition contains optionally nested authority chains back to a root authority 
  - named authorities can be referenced in RULEs
  - the authority is quoted and can be any string
  - this is the CN of the issuer (Root CA) of the client's X.509 certificate

RULEs:
- PROTOCOLS can be 'tls' or 'tcp' for the moment and don't have to be quoted.
  - tls means the connection is TLS.  This can be the case even if client has not provided certificate, and method is not x509 
- METHOD can be 'ca', 'x509', or 'anonymous' and don't have to be quoted
  - x509 means client provided X.509 certificate to attest identity, ca and anonymous are legacy
- AUTHORITY reference must be a defined AUTHORITY definition
- Multiple of same type are combined with OR
- Distinct types are combined with AND

george-mcintyre committed:
Introduce Future Proofing to ACF parsing.

- Allow ACF files to contain sections that are unsupported
- introduce a pragma disable-strict-parsing to disable strict parsing and allow files to be accepted

george-mcintyre committed:
Add tests for future proof parser functionaluty

Expanded tests for Access Security configurations to validate parsing behaviors for unsupported elements. Added detailed cases covering invalid formats, unsupported modifications, and acceptable configurations while maintaining future-proofing and backward compatibility assurances.

george-mcintyre committed:
Refactor parser to handle generic items and floating-point tokens

Replaced "unsupported elements" with a more flexible "generic items" mechanism to enhance parser extensibility. Added support for floating-point numbers in parsing rules. Updated tests and rules to reflect these changes, ensuring better handling of nested and unknown configurations.

george-mcintyre committed:
Add warnings for unsupported ACF elements and update tests

Introduced `yywarn` to log warnings for unsupported ACF elements, rule permissions, and conditions in `asLib.y`. Updated test configuration to replace generic placeholders with more descriptive names. Adjusted test cases and reduced the test plan count to reflect the changes.

george-mcintyre committed:
Remove unused strictParsing logic and improve error handling.

Removed strictParsing variable and related code as it was no longer in use. Enhanced error messaging and parsing logic for handling unsupported elements. Updated test cases for better clarity and expanded coverage of edge cases.

george-mcintyre commented:
This PR includes all commits from the 7.0-future-proof PR #624

george-mcintyre committed:
Fix typo in 'rule_head_mandatory' definition in asLib.y

george-mcintyre committed:
Enhance ASLib parsing with Int64/Float64 support.

Updated ASLib to utilize 64-bit integers and floats for improved parsing precision. Replaced legacy parsing methods with epics-friendly functions, improved error handling, and adjusted tests and grammar accordingly.

george-mcintyre committed:
Refactor AS parser to separate non-rule keywords

Introduced a new `non_rule_keyword` type and moved `tokenASG` and `tokenRULE` under it for better logical grouping. This improves parser clarity and maintains a clean separation of keyword types.

george-mcintyre committed:
Add rule-specific parsing for generic block elements

Introduce new rules for parsing rule-specific generic block elements and their names. This enhances the grammar to handle additional cases while maintaining memory management and error handling. Adjust references to use the new rule where applicable.

george-mcintyre committed:
Refactor grammar rules in asLib.y to remove remaining shift/reduce conflict

Reorganize and refine grammar rules for handling generic elements, blocks, and lists. Simplifies the structure while introducing clearer distinctions between different constructs. This enhances maintainability and removes ambiguities in parsing.

mdavidsaver commented on modules/libcom/src/as/asLib.h:
```diff
@@ -29,6 +29,8 @@ struct dbChannel;
  */
 LIBCOM_API extern int asCheckClientIP;
 
+typedef struct asIdentity *ASIDENTITYPVT;
```
Please add a feature macro to allow API users to clearly identify when the new definitions are present.  eg.

```c
/** Identifies added API associated with asAddClientIdentity()
 * @since UNRELEASED
 */
#define EPICS_ASLIB_HAS_IDENTITY
```

Note that this macro is practically redundant to a test for `S_asLib_dupMethod`.

george-mcintyre replied:
EPICS_ASLIB_HAS_IDENTITY

george-mcintyre committed:
add asLib identity API

george-mcintyre committed:
Use epics types for bytes

mdavidsaver committed:
ascheck cleanup

mdavidsaver committed:
asLib: fix generic lexer/parser

george-mcintyre committed:
Add detailed ACF forward compatibility documentation.

george-mcintyre committed:
Add detailed ACF forward compatibility documentation.

george-mcintyre committed:
Fix typo in PR-624 documentation

george-mcintyre committed:
Simplify memory deallocation by removing redundant casts to `(void *)`.

george-mcintyre committed:
Clarify warnings for unsupported rule keywords in `asLib.y`.

george-mcintyre committed:
Clarify warnings for unsupported rule keywords in `asLib.y`.

george-mcintyre committed:
Add check for empty HAG list in `asLibRoutines.c`.

george-mcintyre committed:
Clean up memory handling and simplify grammar in `asLib.y`

- Adjust `free()` calls to remove unnecessary casting.
- Reorganize `%union` declaration for improved readability.
- Remove unused and redundant rules for unsupported predicates.
- Streamline parsing rules to maintain clarity and consistency.

george-mcintyre committed:
Fix inconsistent indentation in test rules across `aslibtest.c`

- Replaced spaces with tabs for alignment consistency.
- Updated test rules and expected configurations to follow uniform formatting.

george-mcintyre committed:
Remove Docker multi-platform EPICS base build workflow

- Deleted `.github/workflows/docker-pub-epics-base.yml` as the workflow is no longer required.

george-mcintyre committed:
Remove redundant code and streamline comments in `asLib.*`

- Deleted unused function declarations and detailed doxygen-style comments for better readability.
- Updated grammar and memory routines for consolidation in `asLib.y`.

george-mcintyre committed:
"Fix missing and inconsistent indentation, and enhance readability in `aslibtest.c` and `asLib.h`"

- Standardize indentation across test configurations in `aslibtest.c` by replacing tabs with spaces.
- Add a new `MAX_AUTH_NAME_STRING` constant to `asLib.h` for improving code clarity.
- Update formatting in parsing routines and error handling for better maintainability.

george-mcintyre committed:
"Improve readability by adding form feed separators in `asTrapWrite.c`"

george-mcintyre committed:
Update user group assignments in `aslibtest.c` for ORNL, SNS, and HFIR

- Changed user assignments for various administrative and operational groups to reflect updated personnel.
- Adjusted comments in the test data to match the new user assignments for clarity and accuracy.

george-mcintyre committed:
Merge remote-tracking branch 'slac/7.0-future-proof' into 7.0-secure-pvaccess

# Conflicts:
#	modules/libcom/src/as/asLib.h
#	modules/libcom/src/as/asLib.y
#	modules/libcom/src/as/asLibRoutines.c
#	modules/libcom/test/aslibtest.c

george-mcintyre committed:
Enhance EPICS Security with Methods, Authorities, and Protocols

- Added support for METHOD, AUTHORITY, and PROTOCOL tokens in `asLib_lex.l`.
- Extended `asLibRoutines.c` to handle Authorities and Methods in ASG configurations.
- Improved `asLib.y` grammar to incorporate new authority chain and certificate path management.
- Updated comments and documentation to reflect new features.
- Refactored access security functions to support these enhancements, ensuring consistency in handling ASG clients and rules.

george-mcintyre committed:
Replace `strcasecmp` with `epicsStrCaseCmp` in `asLib.y` for protocol name comparison.

george-mcintyre committed:
Use `osiUnistd.h` instead of `unistd.h` in `aslibtest.c` for compatibility.

george-mcintyre committed:
Introduce portable thread-local storage macro and refactor identity initialization in `aslibtest.c`.

george-mcintyre commented:
@george-mcintyre to do a rebase and investigate ci failures and will discuss with @mdavidsaver about how we can incorporate hosts with port numbers into HAGs

mdavidsaver commented:
AI on GM: rebase and investigate CI failures

AI on GM and MD: investigate how HAG entries could specify a port number.
